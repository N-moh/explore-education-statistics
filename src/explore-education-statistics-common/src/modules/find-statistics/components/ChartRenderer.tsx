import HorizontalBarBlock from '@common/modules/find-statistics/components/charts/HorizontalBarBlock';
import LineChartBlock from '@common/modules/find-statistics/components/charts/LineChartBlock';
import { MapFeature } from '@common/modules/find-statistics/components/charts/MapBlock';
import VerticalBarBlock from '@common/modules/find-statistics/components/charts/VerticalBarBlock';
import { Axis, ReferenceLine } from '@common/services/publicationService';
import dynamic from 'next-server/dynamic';
import React from 'react';
import {
  DataBlockData,
  DataBlockMetadata,
} from '@common/services/dataBlockService';

const DynamicMapBlock = dynamic(
  () => import('@common/modules/find-statistics/components/charts/MapBlock'),
  {
    ssr: false,
  },
);

export interface ChartRendererProps {
  type: string;
  indicators: string[];
  data: DataBlockData;
  meta: DataBlockMetadata;
  xAxis: Axis;
  yAxis: Axis;
  height?: number;
  width?: number;
  stacked?: boolean;
  referenceLines?: ReferenceLine[];

  geometry?: MapFeature;
}

function ChartRenderer(props: ChartRendererProps) {
  const {
    data,
    geometry,
    height,
    width,
    meta,
    indicators,
    referenceLines,
    stacked,
    type,
    xAxis = { title: '' },
    yAxis = { title: '' },
  } = props;

  const labels = Object.entries(meta.indicators).reduce(
    (results, [key, indicator]) => ({ ...results, [key]: indicator.label }),
    {},
  );

  // TODO : Temporary sort on the results to get them in date order
  data.result.sort((a, b) => {
    if (a.year < b.year) {
      return -1;
    }

    if (a.year > b.year) {
      return 1;
    }

    return 0;
  });

  switch (type.toLowerCase()) {
    case 'line':
      return (
        <LineChartBlock
          indicators={indicators}
          data={data}
          meta={meta}
          labels={labels}
          xAxis={xAxis}
          yAxis={yAxis}
          height={height}
          width={width}
          referenceLines={referenceLines}
        />
      );
    case 'verticalbar':
      return (
        <VerticalBarBlock
          indicators={indicators}
          data={data}
          meta={meta}
          labels={labels}
          xAxis={xAxis}
          yAxis={yAxis}
          height={height}
          width={width}
          referenceLines={referenceLines}
        />
      );
    case 'horizontalbar':
      return (
        <HorizontalBarBlock
          indicators={indicators}
          data={data}
          meta={meta}
          labels={labels}
          xAxis={xAxis}
          yAxis={yAxis}
          height={height}
          width={width}
          stacked={stacked}
          referenceLines={referenceLines}
        />
      );
    case 'map':
      return (
        <DynamicMapBlock
          indicators={indicators}
          data={data}
          meta={meta}
          labels={labels}
          xAxis={xAxis}
          yAxis={yAxis}
          height={height}
          width={width}
          geometry={geometry}
        />
      );
    default:
      return <div>[ Unimplemented chart type requested ${type} ]</div>;
  }
}

export default ChartRenderer;
