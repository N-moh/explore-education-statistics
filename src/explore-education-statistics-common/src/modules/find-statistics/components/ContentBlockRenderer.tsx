import { ContentBlock } from '@common/services/types/blocks';
import React from 'react';
import ReactMarkdown from 'react-markdown';
import ContentHtml from '@common/components/ContentHtml';

interface Props {
  block: ContentBlock;
}

const ContentBlockRenderer = ({ block }: Props) => {
  const { body = '' } = block;

  switch (block.type) {
    case 'MarkDownBlock':
      return <ReactMarkdown className="dfe-content" source={body} />;
    case 'HtmlBlock':
      return <ContentHtml html={body} />;
    default:
      return null;
  }
};

export default ContentBlockRenderer;
