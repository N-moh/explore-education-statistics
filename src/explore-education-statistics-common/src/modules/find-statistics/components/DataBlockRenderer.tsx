import LoadingSpinner from '@common/components/LoadingSpinner';
import ChartRenderer from '@common/modules/charts/components/ChartRenderer';
import { parseMetaData } from '@common/modules/charts/util/chartUtils';
import dataBlockService, {
  DataBlockResponse,
} from '@common/services/dataBlockService';
import { DataBlock } from '@common/services/types/blocks';
import React, { useEffect, useState } from 'react';
import TimePeriodDataTableRenderer from './TimePeriodDataTableRenderer';

interface DataBlockWithOptionalResponse extends DataBlock {
  dataBlockResponse?: DataBlockResponse;
}

interface Props {
  dataBlock: DataBlockWithOptionalResponse;
  type: 'table' | 'chart';
}

/**
 * Component that takes a {@param dataBlock} (with optional response)
 * and a {@param type} to handle rendering of tables/charts from
 * data block data.
 *
 * The component will handle requesting the data block's
 * response if it was not provided.
 */
const DataBlockRenderer = ({ dataBlock, type }: Props) => {
  const [dataBlockResponse, setDataBlockResponse] = useState<
    DataBlockResponse | undefined
  >(dataBlock.dataBlockResponse);
  const [error, setError] = useState<string>('');

  useEffect(() => {
    if (!dataBlockResponse) {
      dataBlockService
        .getDataBlockForSubject(dataBlock.dataBlockRequest)
        .then(setDataBlockResponse)
        .catch(() => {
          setDataBlockResponse(undefined);
          setError('Unable to get data response from the server');
        });
    }
  }, [dataBlock, dataBlockResponse]);

  if (!dataBlockResponse && error) {
    return <>{error}</>;
  }

  if (!dataBlockResponse) {
    return <LoadingSpinner text="Loading data" />;
  }

  if (type === 'table' && dataBlock.tables.length) {
    return (
      <>
        <TimePeriodDataTableRenderer
          heading={dataBlock.heading}
          response={dataBlockResponse}
          tableHeaders={dataBlock.tables[0].tableHeaders}
        />
      </>
    );
  }

  if (type === 'chart' && dataBlock.charts.length) {
    // There is a presumption that the configuration from the API is valid.
    // The data coming from the API is required to be optional for the ChartRenderer
    // But the data for the charts is required. The charts have validation that
    // prevent them from attempting to render.
    return (
      <ChartRenderer
        {...dataBlock.charts[0]}
        data={dataBlockResponse}
        meta={parseMetaData(dataBlockResponse.metaData)}
      />
    );
  }

  return null;
};

export default DataBlockRenderer;
