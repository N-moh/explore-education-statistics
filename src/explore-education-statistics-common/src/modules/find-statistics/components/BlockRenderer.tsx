import ButtonLink from '@common/components/ButtonLink';
import DataBlock from '@common/modules/find-statistics/components/DataBlock';
import { Publication } from '@common/services/publicationService';
import { Block } from '@common/services/types/blocks';
import React from 'react';
import ReactMarkdown from 'react-markdown';

export type SectionToggleHandler = (section: {
  id: string;
  title: string;
}) => void;

interface Props {
  block?: Block;
  id: string;
  publication?: Publication;
  onToggle?: SectionToggleHandler;
}

const BlockRenderer = ({
  block,
  id,
  publication,
  onToggle,
}: Props) => {
  if (block === undefined) return null;
  switch (block.type) {
    case 'MarkDownBlock':
      return <ReactMarkdown className="govuk-body" source={block.body} />;
    case 'HtmlBlock':
      return (
        <div
          // eslint-disable-next-line react/no-danger
          dangerouslySetInnerHTML={{ __html: block.body }}
        />
      );
    case 'DataBlock':
      return (
        <DataBlock
          {...block}
          id={`${id}_datablock`}
          additionalTabContent={
            <div className="dfe-print-hidden">
              <h2 className="govuk-heading-m govuk-!-margin-top-9">
                Explore and edit this data online
              </h2>
              <p>Use our table tool to explore this data.</p>
              {publication ? (
                <ButtonLink
                  as={`/data-tables/${publication.slug}`}
                  to={`/data-tables?publicationSlug=${publication.slug}`}
                  href={`/data-tables?publicationSlug=${publication.slug}`}
                >
                  Explore data
                </ButtonLink>
              ) : (
                <ButtonLink
                  as="/data-tables"
                  to="/data-tables"
                  href="/data-tables"
                >
                  Explore data
                </ButtonLink>
              )}
            </div>
          }
          onToggle={onToggle}
        />
      );
    default:
      return null;
  }
};

export default BlockRenderer;
