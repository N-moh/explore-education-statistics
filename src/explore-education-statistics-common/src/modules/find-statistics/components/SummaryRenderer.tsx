/* eslint-disable react/no-danger */
import { DetailsToggleHandler } from '@common/components/Details';
import React from 'react';
import { DataBlockData, Result } from '@common/services/dataBlockService';
import { ChartMetaData } from '@common/modules/find-statistics/components/charts/ChartFunctions';
import styles from './SummaryRenderer.module.scss';
import KeyStatTile from './KeyStatTile';

export interface SummaryRendererProps {
  data: DataBlockData;
  meta: ChartMetaData;
  dataKeys: string[];
  dataSummary: string[];
  dataDefinition: string[];
  // description: { type: string; body: string };
  onToggle?: DetailsToggleHandler;
}

export function getLatestMeasures(result: Result[]) {
  const copiedResult = [...result];

  copiedResult.sort((a, b) => a.timePeriod.localeCompare(b.timePeriod));

  const latest = copiedResult[copiedResult.length - 1];

  const { measures } = latest;
  return measures;
}

export default function SummaryRenderer({
  meta,
  // description,
  data,
  dataKeys,
  dataSummary,
  dataDefinition,
  onToggle,
}: SummaryRendererProps) {
  let measures: { [key: string]: string } = {};

  if (meta === undefined) {
    return <div>Unable to render summary, invalid data configured</div>;
  }

  if (data) {
    measures = getLatestMeasures(data.result);
  } else {
    dataKeys.forEach(key => {
      measures[key] = '';
    });
  }

  return (
    <>
      <div className={styles.keyStatsContainer}>
        {dataKeys.map((key, index) => {
          const indicatorKey = `${key}_${index}`;

          return (
            <KeyStatTile
              key={indicatorKey}
              meta={meta}
              measures={measures}
              indicatorKey={key}
              onToggle={onToggle}
              dataDefinition={dataDefinition[index]}
              dataSummary={dataSummary[index]}
            />
          );
        })}
      </div>
      {/* description && description.body !== '' && (
        <ReactMarkdown source={description.body} />
      ) */}
    </>
  );
}
