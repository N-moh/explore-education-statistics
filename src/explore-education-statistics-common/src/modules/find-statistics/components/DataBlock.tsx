/* eslint-disable */
import Tabs from '@common/components/Tabs';
import TabsSection from '@common/components/TabsSection';
import { MapFeature } from '@common/modules/find-statistics/components/charts/MapBlock';
import SummaryRenderer, {
  SummaryRendererProps,
} from '@common/modules/find-statistics/components/SummaryRenderer';
import TableRenderer, {
  Props as TableRendererProps,
} from '@common/modules/find-statistics/components/TableRenderer';
import {
  Chart,
  DataQuery,
  Summary,
  Table,
} from '@common/services/publicationService';

import React, { Component, ReactNode } from 'react';
import DataBlockService, {
  DataBlockRequest,
  DataBlockMetadata,
  DataBlockData,
} from '@common/services/dataBlockService';
import ChartRenderer, {
  ChartRendererProps,
} from '@common/modules/find-statistics/components/ChartRenderer';

export interface DataBlockProps {
  id: string;
  type: string;
  heading?: string;
  dataBlockRequest?: DataBlockRequest;
  charts?: Chart[];
  tables?: Table[];

  summary?: Summary;

  height?: number;
  showTables?: boolean;
  additionalTabContent?: ReactNode;
}

interface DataBlockState {
  charts?: ChartRendererProps[];
  // downloads?: any[];
  tables?: TableRendererProps[];
  summary?: SummaryRendererProps;
}

class DataBlock extends Component<DataBlockProps, DataBlockState> {
  public static defaultProps = {
    showTables: true,
  };

  public state: DataBlockState = {};

  private currentDataQuery?: DataQuery = undefined;

  public async componentDidMount() {
    const { dataBlockRequest } = this.props;

    if (dataBlockRequest) {
      const result = await DataBlockService.getDataBlockForSubject(
        dataBlockRequest,
      );

      this.parseDataResponse(result.data, result.metaData);
    } else {
      // this.parseDataResponse(data, meta);
    }
  }

  public async componentWillUnmount() {
    this.currentDataQuery = undefined;
  }

  private parseDataResponse(
    json: DataBlockData,
    jsonMeta: DataBlockMetadata,
  ): void {
    const newState: DataBlockState = {};

    const { charts, summary, tables } = this.props;

    if (json.result.length > 0) {
      if (tables) {
        newState.tables = [{ data: json, meta: jsonMeta, ...tables[0] }];
      } else {
        //TODO: remove when data is updated
        newState.tables = [
          {
            data: json,
            meta: jsonMeta,
            indicators: Object.keys(json.result[0].measures),
          },
        ];
      }
    }

    if (charts) {
      newState.charts = charts.map(chart => ({
        xAxis: { title: '' },
        yAxis: { title: '' },

        ...chart,

        data: json,
        meta: jsonMeta,
      }));
    }

    if (summary) {
      newState.summary = {
        ...summary,
        data: json,
        meta: jsonMeta,
      };
    }
    this.setState(newState);
  }

  public render() {
    const {
      heading = '',
      height,
      showTables,
      additionalTabContent,
      id,
    } = this.props;
    const { charts, summary, tables } = this.state;

    return (
      <div className="govuk-datablock" data-testid={`DataBlock ${heading}`}>
        <Tabs>
          {summary && (
            <TabsSection id={`datablock_${id}_summary`} title="Summary">
              <h3>{heading}</h3>
              <SummaryRenderer {...summary} />
            </TabsSection>
          )}

          {tables && showTables && (
            <TabsSection id={`datablock_${id}_tables`} title="Data tables">
              <h3>{heading}</h3>
              {tables.map((table, idx) => {
                const key = `${id}0_table_${idx}`;

                return <TableRenderer key={key} {...table} />;
              })}
              {additionalTabContent}
            </TabsSection>
          )}

          {charts && (
            <TabsSection
              id={`datablock_${id}_charts`}
              title="Charts"
              lazy={false}
            >
              <h3>{heading}</h3>
              {charts.map((chart, idx) => {
                const key = `${id}_chart_${idx}`;

                return <ChartRenderer key={key} {...chart} height={height} />;
              })}
              {additionalTabContent}
            </TabsSection>
          )}

          <TabsSection id={`datablock_${id}_downloads`} title="Data downloads">
            <p>
              You can customise and download data as Excel, .csv or .pdf files.
              Our data can also be accessed via an API.
            </p>
            <div className="govuk-inset-text">
              Data downloads have not yet been implemented within the service.
            </div>
            {additionalTabContent}
          </TabsSection>
        </Tabs>
      </div>
    );
  }
}

export default DataBlock;
