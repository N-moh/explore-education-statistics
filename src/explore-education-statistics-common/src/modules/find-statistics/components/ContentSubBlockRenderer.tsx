import DataBlock from '@common/modules/find-statistics/components/DataBlock';
import { ContentBlock, Publication } from '@common/services/publicationService';
import ButtonLink from '@common/components/ButtonLink';
import React from 'react';
import ReactMarkdown from 'react-markdown';

export type SectionToggleHandler = (section: {
  id: string;
  title: string;
}) => void;

interface Props {
  block?: ContentBlock;
  id: string;
  publication: Publication;
  onToggle?: SectionToggleHandler;
}

const ContentSubBlockRenderer = ({
  block,
  id,
  publication,
  onToggle,
}: Props) => {
  if (block === undefined) return null;
  switch (block.type) {
    case 'MarkDownBlock':
      return <ReactMarkdown className="govuk-body" source={block.body} />;
    case 'HtmlBlock':
      return (
        <div
          // eslint-disable-next-line react/no-danger
          dangerouslySetInnerHTML={{ __html: block.body }}
        />
      );
    case 'InsetTextBlock':
      return (
        <div className="govuk-inset-text">
          {block.heading && (
            <h3 className="govuk-heading-s">{block.heading}</h3>
          )}

          <ReactMarkdown className="govuk-body" source={block.body} />
        </div>
      );
    case 'DataBlock':
      return (
        <DataBlock {...block} id={`${id}_datablock`} onToggle={onToggle} />
      );
    default:
      return null;
  }
};

export default ContentSubBlockRenderer;
