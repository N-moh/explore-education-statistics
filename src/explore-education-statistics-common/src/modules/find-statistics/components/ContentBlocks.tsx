import { Publication, Release } from '@common/services/publicationService';
import React from 'react';
import ContentSubBlockRenderer, {
  SectionToggleHandler,
} from './ContentSubBlockRenderer';

export interface ContentBlockProps {
  content: Release['content'][0]['content'];
  id: string;
  publication: Publication;
  onToggle?: SectionToggleHandler;
}

const ContentBlocks = ({
  content,
  id,
  publication,
  onToggle,
}: ContentBlockProps) => {
  return content && content.length > 0 ? (
    <>
      {content.map((block, index) => {
        const key = `${index}-${block.heading}-${block.type}`;
        return (
          <ContentSubBlockRenderer
            id={id}
            block={block}
            key={key}
            publication={publication}
            onToggle={onToggle}
          />
        );
      })}
    </>
  ) : (
    <div className="govuk-inset-text">
      There is no content for this section.
    </div>
  );
};

export default ContentBlocks;
