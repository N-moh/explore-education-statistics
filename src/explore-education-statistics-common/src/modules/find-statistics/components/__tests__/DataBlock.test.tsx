import testData from '@common/modules/find-statistics/components/charts/__tests__/__data__/testBlockData';
import _dataBlockService, {
  DataBlockRequest,
  GeographicLevel,
} from '@common/services/dataBlockService';
import React from 'react';
import { render, wait } from 'react-testing-library';
import DataBlock from '../DataBlock';
import { SummaryRendererProps } from '@common/modules/find-statistics/components/SummaryRenderer';

jest.mock('@common/services/dataBlockService');

const dataBlockService = _dataBlockService as jest.Mocked<
  typeof _dataBlockService
>;

describe('DataBlock', () => {
  const dataBlockRequest: DataBlockRequest = {
    subjectId: 1,
    geographicLevel: GeographicLevel.National,
    startYear: '2014',
    endYear: '2015',
    filters: ['1', '2'],
    indicators: ['23', '26', '28'],
  };

  test('renders horizontal chart', async () => {
    const getDataBlockForSubject = dataBlockService.getDataBlockForSubject.mockImplementation(
      (_: DataBlockRequest) => {
        return Promise.resolve(testData.response);
      },
    );

    const { container } = render(
      <DataBlock
        id="test"
        type="datablock"
        dataBlockRequest={dataBlockRequest}
        showTables={false}
        charts={[
          {
            type: 'horizontalbar',
            stacked: true,
            indicators: ['23', '26', '28'],
            xAxis: { title: 'test x axis' },
            yAxis: { title: 'test y axis' },
            width: 800,
            height: 600,
          },
        ]}
      />,
    );

    await wait();

    expect(getDataBlockForSubject).toBeCalledWith(dataBlockRequest);

    expect(
      container.querySelectorAll('section.govuk-tabs__panel'),
    ).toHaveLength(1);

    expect(
      container.querySelector('section.govuk-tabs__panel h3'),
    ).toHaveTextContent('Charts');

    expect(container.querySelectorAll('.recharts-bar')).toHaveLength(3);
  });

  test('renders vertical chart', async () => {
    const getDataBlockForSubject = dataBlockService.getDataBlockForSubject.mockImplementation(
      (_: DataBlockRequest) => {
        return Promise.resolve(testData.response);
      },
    );

    const { container } = render(
      <DataBlock
        id="test"
        type="datablock"
        dataBlockRequest={dataBlockRequest}
        showTables={false}
        charts={[
          {
            type: 'verticalbar',
            indicators: ['23', '26', '28'],
            xAxis: { title: 'test x axis' },
            yAxis: { title: 'test y axis' },
            width: 800,
            height: 600,
          },
        ]}
      />,
    );

    await wait();

    expect(getDataBlockForSubject).toBeCalledWith(dataBlockRequest);

    expect(
      container.querySelectorAll('section.govuk-tabs__panel'),
    ).toHaveLength(1);

    expect(
      container.querySelector('section.govuk-tabs__panel h3'),
    ).toHaveTextContent('Charts');

    expect(container.querySelectorAll('.recharts-bar')).toHaveLength(3);
  });

  test('renders table', async () => {
    const getDataBlockForSubject = dataBlockService.getDataBlockForSubject.mockImplementation(
      (_: DataBlockRequest) => {
        return Promise.resolve(testData.response);
      },
    );

    const { container } = render(
      <DataBlock
        id="test"
        type="datablock"
        dataBlockRequest={dataBlockRequest}
        showTables
      />,
    );

    await wait();

    expect(getDataBlockForSubject).toBeCalledWith(dataBlockRequest);

    expect(container.querySelector('table')).toMatchSnapshot();
  });

  test('renders summary', async () => {
    const getDataBlockForSubject = dataBlockService.getDataBlockForSubject.mockImplementation(
      (_: DataBlockRequest) => {
        return Promise.resolve(testData.response);
      },
    );

    const { container } = render(
      <DataBlock
        id="test"
        type="databock"
        dataBlockRequest={dataBlockRequest}
        showTables={false}
        // eslint-disable-next-line @typescript-eslint/no-object-literal-type-assertion
        summary={
          {
            dataKeys: ['23', '26', '28'],
            dataSummary: ['up 10%', 'down 10%', 'up 11%'],
            description: {
              type: 'MarkDownBlock',
              body: `<div>test</div>`,
            },
          } as SummaryRendererProps
        }
      />,
    );

    await wait();

    expect(getDataBlockForSubject).toBeCalledWith(dataBlockRequest);

    expect(
      container.querySelector('#datablock_test_summary'),
    ).toMatchSnapshot();
  });
});
