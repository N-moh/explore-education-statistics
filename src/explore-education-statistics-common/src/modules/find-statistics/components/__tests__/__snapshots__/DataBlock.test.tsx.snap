// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataBlock datablock renders map 1`] = `

<div data-testid="DataBlock ">
  <div class="govuk-tabs">
    <ul class="govuk-tabs__list"
        role="tablist"
    >
      <li class="govuk-tabs__list-item"
          role="presentation"
      >
        <a class="govuk-tabs__tab govuk-tabs__tab--selected"
           href="#datablock_test_charts"
           id="datablock_test_charts-tab"
           aria-controls="datablock_test_charts"
           aria-selected="true"
           role="tab"
        >
          Charts
        </a>
      </li>
    </ul>
    <section class="govuk-tabs__panel dfe-content-overflow panel"
             id="datablock_test_charts"
             data-testid="Charts"
             aria-labelledby="datablock_test_charts-tab"
             role="tabpanel"
             tabindex="-1"
    >
      <h3>
        Charts
      </h3>
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-one-third"
             aria-live="assertive"
        >
          <form>
            <div class="govuk-form-group govuk-!-margin-bottom-6">
              <label class="govuk-label"
                     for="selectedIndicator"
              >
                Select data to view
              </label>
              <select class="govuk-select select"
                      id="selectedIndicator"
                      name="selectedIndicator"
              >
                <option value="23">
                  Unauthorised absence rate
                </option>
                <option value="26">
                  Overall absence rate
                </option>
                <option value="28">
                  Authorised absence rate
                </option>
              </select>
            </div>
            <div class="govuk-form-group govuk-!-margin-bottom-6">
              <label class="govuk-label"
                     for="selectedLocation"
              >
                Select a location
              </label>
              <select class="govuk-select select"
                      id="selectedLocation"
                      name="selectedLocation"
              >
                <option value>
                  select...
                </option>
                <option value="E92000001">
                  England
                </option>
              </select>
            </div>
          </form>
          <div>
            <h3 class="govuk-heading-s">
              Key to Authorised absence rate
            </h3>
            <dl class="govuk-list">
              <dd class="legend">
                <span class="rate0">
                  &nbsp;
                </span>
                5.0%&nbsp; to 5.0%
              </dd>
              <dd class="legend">
                <span class="rate1">
                  &nbsp;
                </span>
                5.0%&nbsp; to 5.0%
              </dd>
              <dd class="legend">
                <span class="rate2">
                  &nbsp;
                </span>
                5.0%&nbsp; to 5.0%
              </dd>
              <dd class="legend">
                <span class="rate3">
                  &nbsp;
                </span>
                5.0%&nbsp; to 5.0%
              </dd>
              <dd class="legend">
                <span class="rate4">
                  &nbsp;
                </span>
                5.0%&nbsp; to 5.0%
              </dd>
            </dl>
          </div>
        </div>
        <div class="govuk-grid-column-two-thirds">
          <div class="map leaflet-container leaflet-grab leaflet-touch-drag"
               style="width: 800px; height: 600px; position: relative;"
               tabindex="0"
          >
            <div class="leaflet-pane leaflet-map-pane"
                 style="left: 0px; top: 0px;"
            >
              <div class="leaflet-pane leaflet-tile-pane">
              </div>
              <div class="leaflet-pane leaflet-shadow-pane">
              </div>
              <div class="leaflet-pane leaflet-overlay-pane">
                <svg pointer-events="none"
                     width="0"
                     height="0"
                     style="left: 0px; top: 0px;"
                     viewbox="0 0 0 0"
                >
                  <g>
                    <path class="uk leaflet-interactive"
                          stroke="#3388ff"
                          stroke-opacity="1"
                          stroke-width="3"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          fill="#3388ff"
                          fill-opacity="0.2"
                          fill-rule="evenodd"
                          d="M-3 3L-3 -3L3 -3L3 3z"
                    >
                    </path>
                    <path class="rateNaN leaflet-interactive"
                          stroke="#3388ff"
                          stroke-opacity="1"
                          stroke-width="3"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          fill="#3388ff"
                          fill-opacity="0.2"
                          fill-rule="evenodd"
                          d="M0 0"
                    >
                    </path>
                  </g>
                </svg>
              </div>
              <div class="leaflet-pane leaflet-marker-pane">
              </div>
              <div class="leaflet-pane leaflet-tooltip-pane">
              </div>
              <div class="leaflet-pane leaflet-popup-pane">
              </div>
            </div>
            <div class="leaflet-control-container">
              <div class="leaflet-top leaflet-left">
                <div class="leaflet-control-zoom leaflet-bar leaflet-control">
                  <a class="leaflet-control-zoom-in"
                     href="#"
                     title="Zoom in"
                     role="button"
                     aria-label="Zoom in"
                  >
                    +
                  </a>
                  <a class="leaflet-control-zoom-out"
                     href="#"
                     title="Zoom out"
                     role="button"
                     aria-label="Zoom out"
                  >
                    âˆ’
                  </a>
                </div>
              </div>
              <div class="leaflet-top leaflet-right">
              </div>
              <div class="leaflet-bottom leaflet-left">
              </div>
              <div class="leaflet-bottom leaflet-right">
                <div class="leaflet-control-attribution leaflet-control">
                  <a href="https://leafletjs.com"
                     title="A JS library for interactive maps"
                  >
                    Leaflet
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>
</div>

`;

exports[`DataBlock renders summary 1`] = `
<section
  aria-labelledby="datablock_test_summary-tab"
  class="govuk-tabs__panel dfe-content-overflow panel"
  data-testid="Summary"
  id="datablock_test_summary"
  role="tabpanel"
  tabindex="-1"
>
  <h3>
    Summary
  </h3>
  <div
    class="keyStatsContainer"
  >
    <div
      class="keyStatTile"
    >
      <div
        class="keyStat"
      >
        <h3
          class="govuk-heading-s"
        >
          Unauthorised absence rate
        </h3>
        <p
          class="govuk-heading-xl"
          data-testid="tile Unauthorised absence rate"
        >
          3
          %
        </p>
        <p
          class="govuk-body-s"
        >
          up 10%
        </p>
      </div>
      <details
        class="govuk-details"
        role="group"
      >
        <summary
          aria-controls="details-content-4"
          aria-expanded="false"
          class="govuk-details__summary"
          role="button"
          tabindex="0"
        >
          <span
            class="govuk-details__summary-text"
            data-testid="details--expand"
          >
            What is Unauthorised absence rate?
          </span>
        </summary>
        <div
          aria-hidden="true"
          class="govuk-details__text"
          id="details-content-4"
        >
          Overall absence is the adipisicing elit. Dolorum hic nobis voluptas quidem fugiat enim ipsa reprehenderit nulla.
        </div>
      </details>
    </div>
    <div
      class="keyStatTile"
    >
      <div
        class="keyStat"
      >
        <h3
          class="govuk-heading-s"
        >
          Overall absence rate
        </h3>
        <p
          class="govuk-heading-xl"
          data-testid="tile Overall absence rate"
        >
          10
          %
        </p>
        <p
          class="govuk-body-s"
        >
          down 10%
        </p>
      </div>
      <details
        class="govuk-details"
        role="group"
      >
        <summary
          aria-controls="details-content-5"
          aria-expanded="false"
          class="govuk-details__summary"
          role="button"
          tabindex="0"
        >
          <span
            class="govuk-details__summary-text"
            data-testid="details--expand"
          >
            What is Overall absence rate?
          </span>
        </summary>
        <div
          aria-hidden="true"
          class="govuk-details__text"
          id="details-content-5"
        >
          Overall absence is the adipisicing elit. Dolorum hic nobis voluptas quidem fugiat enim ipsa reprehenderit nulla.
        </div>
      </details>
    </div>
    <div
      class="keyStatTile"
    >
      <div
        class="keyStat"
      >
        <h3
          class="govuk-heading-s"
        >
          Authorised absence rate
        </h3>
        <p
          class="govuk-heading-xl"
          data-testid="tile Authorised absence rate"
        >
          5
          %
        </p>
        <p
          class="govuk-body-s"
        >
          up 11%
        </p>
      </div>
      <details
        class="govuk-details"
        role="group"
      >
        <summary
          aria-controls="details-content-6"
          aria-expanded="false"
          class="govuk-details__summary"
          role="button"
          tabindex="0"
        >
          <span
            class="govuk-details__summary-text"
            data-testid="details--expand"
          >
            What is Authorised absence rate?
          </span>
        </summary>
        <div
          aria-hidden="true"
          class="govuk-details__text"
          id="details-content-6"
        >
          Overall absence is the adipisicing elit. Dolorum hic nobis voluptas quidem fugiat enim ipsa reprehenderit nulla.
        </div>
      </details>
    </div>
  </div>
  &lt;div&gt;test&lt;/div&gt;
</section>
`;

exports[`DataBlock renders table 1`] = `null`;
